### PR TITLE
[AutoDiff] Register VJPs for SIMD `subscript(_: Int)` setters.

### PR DESCRIPTION
**Pull request by dan-zheng**

### CODE REVIEW

File: stdlib/public/Differentiation/SIMDDifferentiation.swift.gyb [hunk 1]
```diff
@@ -61,6 +61,25 @@ where
   }
 }
 
+extension SIMD${n}
```
Comment (rxwei) [inline stdlib/public/Differentiation/SIMDDifferentiation.swift.gyb:64]: Why is it not in the same extension as the VJP for the getter? And why do the getter VJP and the setter VJP need different generic constraints?

Comment (dan-zheng) [reply stdlib/public/Differentiation/SIMDDifferentiation.swift.gyb:64]: Thanks for pointing this out.

`SIMD` types currently conditionally conform to `Differentiable` where `Scalar: Differentiable & BinaryFloatingPoint, Scalar.TangentVector: BinaryFloatingPoint`, and have `TangentVector == Self`.

The subscript getter VJP has those same generic requirements.

Since `TangentVector == Self`, the subscript setter VJP additionally needs `Scalar.TangentVector == Scalar` so that the pullback body can type-check: `dSelf[index]` has type `Scalar`, which must be equal to `Scalar.TangentVector`.

---

We could consider using the stronger differentiability generic requirements (`Scalar.TangentVector == Scalar`) everywhere.

Comment (rxwei) [reply stdlib/public/Differentiation/SIMDDifferentiation.swift.gyb:64]: I understand why `Scalar.TangentVector == Scalar` could be a better generic constraint than `Scalar.TangentVector: BinaryFloatingPoint`, but I don't quite see why it's necessary for `_vjpSubscript` and `_vjpSubscriptSetter` to be constrained differently. Note that in `_vjpSubscript`'s pullback the type of `v` and the type of `zeros[index]` are _already different_. We invoked `Scalar.init` on `v` to convert it to a type compatible with `zeros[index]`. In `_vjpSubscriptSetter`, it appears to me like the exact same issue, where I think you can write 
```swift
return .init(dNewValue)
```
... to get it to compile while the generic constraint is `Scalar.TangentVector: BinaryFloatingPoint`.

In short, I feel these two VJPs should have the same generic constraints and be placed in the same extension: either `Scalar.TangentVector == Scalar`, or `Scalar.TangentVector == Scalar`.

Comment (dan-zheng) [reply stdlib/public/Differentiation/SIMDDifferentiation.swift.gyb]: Ah, I missed those details, thanks. I moved both VJPs to the same extension.

**Review by stephentyrone — APPROVED** [review]

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

**Comment by dan-zheng** [comment]
https://github.com/apple/swift/pull/32839 fixed a legitimate test failure. Retriggering tests now.

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test
